### PR TITLE
feat(ci): add PR-size governance gate

### PR DESCRIPTION
**Pull request by qnbs**
## Summary

Adds a PR-size governance gate, closing the last of the three residual clusters (S4, S5, PR-size governance) identified in the post-#477 reconstruction reconciliation program's plan.

## What it checks

`scripts/check-pr-size.mjs` evaluates a PR's diff against tiered file/line/commit limits:

| Tier | Files | Lines | Commits | Applies to |
|---|---|---|---|---|
| target | ≤8 | ≤400 | ≤6 | baseline, every profile |
| hard | ≤20 | ≤1200 | ≤10 | normal (non-docs) profile |
| docsGovernance | ≤15 | ≤2400 | ≤8 | all-DOCS profile (replaces hard, not target) |
| absolute | ≤30 | ≤3000 | ≤15 | **blocking**, regardless of profile |

"Meaningful lines" reuses `classifyFile()` from `scripts/ci-prepush-classifier.mjs` to zero out `NON_CODE_ONLY` diffs (e.g. a locale bundle rebuild) and `pnpm-lock.yaml`, so generated churn can't trip the gate.

This is **not** a restatement of CLAUDE.md's existing "~100 changed files" rule — that number is specifically about CodeAnt AI skipping inline review above 100 files; this gate is a different, stricter, purpose-built size discipline.

## Wiring

- `pnpm run pr-size:check` (new script)
- New `pr-size` CI job, `pull_request`-triggered only, job-level `permissions: contents: read, pull-requests: write`
- **Self-defeat-proofing (partial)**: the job checks out and runs the **base ref's own copy** of `check-pr-size.mjs` (and its `ci-prepush-classifier.mjs` dependency), never the PR's own working-tree copy — this prevents the PR from changing the checker logic or thresholds used for its own evaluation. Same discipline as #505's `workflow-policy` job. **This does not constitute a complete trusted-execution boundary for the PR-controlled workflow job itself** (the `ci.yml` job body and its `pull-requests: write` token still run from the PR's own checkout) — that residual trust-boundary question is tracked separately as a follow-up, not solved here.
- Posts a PR comment only when there's something to report (silent when within target)
- The job's own exit code is 0 except at the absolute tier — folding it into the `ci-success` required aggregator (with the same skipped-is-OK tolerance as `rust-tauri`/`core-rust` for non-`pull_request` events) naturally makes it advisory below absolute and blocking only there, with no separate workflow-graph split needed
- Added `pr-size`'s `pull-requests: write` to `workflow-policy-check.mjs`'s own `WRITE_SCOPE_ALLOWLIST` — verified structurally sound by that same checker

## Scope freeze / deferred follow-up

This PR is under terminal convergence: one final bounded correction batch (nonzero-exit handling, generated-artifact file counting, comment-ID reconciliation, CI-diagram/doc truth, bootstrap-fallback documentation), no new architecture. A P1 finding surfaced during review — the `pr-size` (and `workflow-policy`) job's write-scoped enforcement step still runs from the PR's own checkout, so `needs:` alone doesn't establish a trusted-execution boundary — is real but out of scope here; a trusted-execution redesign does not belong in a PR about bounding *PR size*. Tracked in #510, related to #506 Finding C.

## Test plan

- [x] 23 new unit tests (`tests/unit/tooling/checkPrSize.test.ts`) — DI-injected git output, fail-closed spawn-error handling, tier-boundary cases, docsGovernance-vs-hard profile comparison
- [x] `pnpm run workflow-policy:check` — passes against the updated `ci.yml` (new job, new allowlist entry)
- [x] `pnpm run lint` — clean
- [x] `pnpm exec tsgo --project tsconfig.tsgo.json --noEmit --checkers 4` — clean (exact CI command)
- [x] `pnpm run ci:prepush` — full local admission green
- [x] `git diff --check` — clean
- [x] Fixed a stale hardcoded `ci-success.needs` array in the pre-existing `tests/unit/workflowPolicy.test.ts` (same pattern as #505's `workflow-policy` job addition)
- [x] Self-tested against two real merged-PR ranges (small S5 PR → `ok` tier; large multi-wave S4 PR → correctly flagged `hard` tier)
- [x] Commits SSH-signed

## Summary by Sourcery

Introduce pull-request size governance that advises on oversized changes and blocks pull requests exceeding the absolute limits.

New Features:
- Add tiered pull-request size governance based on changed files, meaningful lines, and commit count, with dedicated limits for documentation-only changes.
- Report oversized pull requests through advisory comments while blocking changes that exceed the absolute ceiling.

Bug Fixes:
- Exclude generated locale bundles, the lockfile, and mirrored template artifacts from governed size metrics so generated churn does not distort enforcement.
- Ensure checker failures and unexpected nonzero exits fail closed, while preserving binary and renamed-file handling.

Enhancements:
- Run the size checker from the base branch when available and integrate its pull-request-only result into the required CI success aggregator.
- Add a local PR-size check command and update CI documentation, workflow diagrams, and policy allowlists.

CI:
- Add a pull_request-only PR-size governance job with controlled pull-request comment reporting and tier-aware enforcement.

Documentation:
- Document the new PR-size governance job, thresholds, CI dependencies, and required-status behavior.

Tests:
- Add comprehensive unit and integration coverage for parsing, generated-file exclusions, tier selection, reporting, failure handling, and attribute-based diff evasion.

Chores:
- Refresh documented test counts and update workflow-policy expectations for the new CI dependency.

___

## **CodeAnt-AI Description**
Add PR size checks to CI with advisory and blocking limits

### What Changed
- Pull requests are measured by changed files, meaningful lines, and commit count against defined size tiers
- Oversized but reviewable PRs receive an advisory comment, while PRs exceeding the absolute ceiling block CI
- Documentation-only PRs use dedicated limits, and generated locale files and lockfile changes do not inflate meaningful line counts
- The check runs using the base branch’s rules and includes coverage for binary files, size profiles, reporting, and failure handling

### Impact
`✅ Fewer oversized pull requests`
`✅ Clearer PR splitting guidance`
`✅ Generated-file churn no longer triggers size warnings`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added automated pull-request size checks with advisory and blocking thresholds.
  * Reports oversized changes directly in pull requests and replaces notices when changes return within target.
  * Documentation-only changes use dedicated size thresholds.

* **CI**
  * Integrated pull-request size results into overall CI status reporting.
  * Added safeguards for bootstrap pull requests and workflow policy validation.

* **Documentation**
  * Updated project test-count metrics and CI workflow guidance.

* **Tests**
  * Added coverage for size calculations, file handling, reporting, and failure scenarios.

### CODE REVIEW

File: scripts/check-pr-size.mjs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+function runGit(args, dependencies = {}) {
+  const spawn = dependencies.spawnSync ?? spawnSync;
+  const result = spawn('git', args, { encoding: 'utf8' });
+  if (result.error || result.status !== 0) return null;
+  return result.stdout;
+}
+
+export function getChangedFilesNumstat(base, head, dependencies = {}) {
+  return runGit(['diff', '--no-renames', '--numstat', `${base}...${head}`], dependencies);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in d0daab1f: switched to git diff --numstat -z, which preserves rename detection — a rename record is 3 NUL-separated tokens (numbers, old path, new path), not the delete+add doubling --no-renames caused. parseNumstat() rewritten for the NUL-delimited format.

File: scripts/check-pr-size.mjs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  return Number.isFinite(count) ? count : null;
+}
+
+// QNBS-v3: binary files report "-\t-\tpath" in numstat — treated as 0 meaningful lines, not NaN.
+export function parseNumstat(numstatOutput) {
+  const rows = [];
+  for (const line of numstatOutput.split('\n')) {
+    if (!line.trim()) continue;
+    const [added, removed, ...pathParts] = line.split('\t');
+    const path = pathParts.join('\t');
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in d0daab1f: -z gives raw UTF-8 paths (no octal-quoting for non-ASCII filenames), same fix as the rename-detection change — both came from switching --no-renames to -z.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -159,6 +159,55 @@ jobs:
...
+
+      # QNBS-v3: stays silent when within target — only comments when there's something to say.
+      - name: Post PR size comment (advisory or blocking tiers only)
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          if grep -qE 'within target' /tmp/pr-size-report.txt; then
+            exit 0
+          fi
+          gh pr comment "${{ github.event.pull_request.number }}" --body-file /tmp/pr-size-report.txt
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in d0daab1f: added continue-on-error: true to the comment-posting step, so a read-only fork-PR token failing gh pr comment can never fail this advisory job. Same fix as the equivalent chatgpt-codex-connector finding on this same line.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -159,6 +159,55 @@ jobs:
...
+      - name: Evaluate PR size (base-ref checker)
+        id: size
+        env:
+          GH_TOKEN: ${{ github.token }}
+          BASE_SHA: ${{ github.event.pull_request.base.sha }}
+          HEAD_SHA: ${{ github.event.pull_request.head.sha }}
+        run: |
+          set -euo pipefail
+          mkdir -p /tmp/base-scripts
+          git show "$BASE_SHA:scripts/check-pr-size.mjs" > /tmp/base-scripts/check-pr-size.mjs
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Confirmed stale — already fixed in the prior commit (a0abfad5), before this review ran. Current ci.yml has the bootstrap fallback (checks base-ref existence via 2>/dev/null, falls back to the PR's own copy only when the checker doesn't exist on base yet). No new fix needed.

File: .github/workflows/ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in d0daab1f: added continue-on-error: true to the comment-posting step (same fix as the CodeRabbit finding on this same line).

File: scripts/check-pr-size.mjs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  }
+  return rows;
+}
+
+// QNBS-v3: zeroes generated/non-code diffs so e.g. a locale bundle rebuild can't trip this gate.
+export function computeMeaningfulLines(rows) {
+  let total = 0;
+  for (const row of rows) {
+    if (row.path.split('/').pop() === 'pnpm-lock.yaml') continue;
+    if (classifyFile(row.path) === 'NON_CODE_ONLY') continue;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in d0daab1f: replaced the blanket NON_CODE_ONLY classifier exemption with a targeted GENERATED_ARTIFACT_ROOTS check (public/locales/, public/community-templates/ only) — locales/**/*.json source content now counts as meaningful lines again.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -492,6 +542,9 @@ jobs:
         run: |
           FAIL=0
           [ "${{ needs.workflow-policy.result }}" = "success" ] || FAIL=1
+          if [ "${{ needs.pr-size.result }}" != "success" ] && [ "${{ needs.pr-size.result }}" != "skipped" ]; then
+            FAIL=1
+          fi
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in d0daab1f: ci-success's pr-size check is now event-aware — 'skipped' is only tolerated when github.event_name != 'pull_request', since pr-size's own if: is exactly that same check and a tampered if: making it skip on a genuine PR would otherwise silently pass.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -159,6 +159,55 @@ jobs:
             echo "crates=false" >> "$GITHUB_OUTPUT"
           fi
 
+  # ----------------------------------------------------------
+  # PR SIZE GOVERNANCE: advisory below the absolute ceiling, blocking above it.
+  # QNBS-v3: runs the BASE ref's checker copy — a PR touching it can't raise its own limits.
+  # ----------------------------------------------------------
+  pr-size:
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml:166]: Fixed in d0daab1f: added needs: [workflow-policy] to the pr-size job, same gating discipline as security's own needs: [workflow-policy].

File: scripts/check-pr-size.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in d0daab1f: same fix as the CodeAnt finding on this line — switched to git diff --numstat -z, which preserves rename detection instead of --no-renames' delete+add doubling.

File: .github/workflows/ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in d0daab1f: the comment-posting step now upserts one comment via --edit-last (stable bot identity) instead of posting a new one on every oversized push, and posts a 'resolved' update once the PR is back within target instead of leaving a stale warning.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -159,6 +159,71 @@ jobs:
...
+      - name: Post PR size comment (advisory or blocking tiers only)
+        continue-on-error: true
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          PR="${{ github.event.pull_request.number }}"
+          MARKER="<!-- pr-size-governance -->"
+          if grep -qE 'within target' /tmp/pr-size-report.txt; then
+            printf '%s\nPR size is back within target — previous warning below is resolved.\n' "$MARKER" > /tmp/pr-size-resolved.txt
+            gh pr comment "$PR" --edit-last --body-file /tmp/pr-size-resolved.txt || true
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in f7a863cf: the step now fetches the last github-actions[bot] comment first and only uses --edit-last when it actually contains the  marker; otherwise it posts a fresh comment instead of overwriting an unrelated job's comment.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -159,6 +159,71 @@ jobs:
...
           fi
 
+  # ----------------------------------------------------------
+  # PR SIZE GOVERNANCE: advisory below the absolute ceiling, blocking above it.
+  # QNBS-v3: runs the BASE ref's checker copy — a PR touching it can't raise its own limits.
+  # ----------------------------------------------------------
+  pr-size:
+    name: 📏 PR Size Governance
+    if: github.event_name == 'pull_request'
+    needs: [workflow-policy]
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml:169]: Fixed in f7a863cf: both AGENTS.md's Pipeline Graph/Jobs table and docs/CI.md's job graph/gate-authority list/job table now include pr-size, its needs: workflow-policy dependency, and its pull_request-only skip condition.

File: scripts/check-pr-size.mjs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+function runGit(args, dependencies = {}) {
+  const spawn = dependencies.spawnSync ?? spawnSync;
+  const result = spawn('git', args, { encoding: 'utf8' });
+  if (result.error || result.status !== 0) return null;
+  return result.stdout;
+}
+
+// QNBS-v3: -z gives raw UTF-8 paths (git otherwise octal-escapes non-ASCII) and keeps rename detection.
+export function getChangedFilesNumstat(base, head, dependencies = {}) {
+  return runGit(['diff', '--numstat', '-z', `${base}...${head}`], dependencies);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in f7a863cf: writes a temporary '* diff' override to $GIT_DIR/info/attributes (resolved via git rev-parse --git-path info/attributes, worktree-aware) before running the diff, then restores the original content. That file is never part of the versioned tree, so a PR can't touch it — confirmed --text/-a and -c core.attributesFile=/dev/null do NOT override an explicit -diff attribute, only this approach does. New integration test with a real temporary git repo proves both the vulnerability and the fix.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -159,6 +159,80 @@ jobs:
...
+              printf '%s\nPR size is back within target — previous warning below is resolved.\n' "$MARKER" > /tmp/pr-size-resolved.txt
+              gh pr comment "$PR" --edit-last --body-file /tmp/pr-size-resolved.txt || true
+            fi
+            exit 0
+          fi
+          { printf '%s\n' "$MARKER"; cat /tmp/pr-size-report.txt; } > /tmp/pr-size-marked.txt
+          if [ "$LAST_IS_OURS" = "true" ]; then
+            gh pr comment "$PR" --edit-last --body-file /tmp/pr-size-marked.txt || true
+          else
+            gh pr comment "$PR" --body-file /tmp/pr-size-marked.txt || true
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in ea58ddbc — the upsert now searches all `github-actions[bot]` comments for the `<!-- pr-size-governance -->` marker (not just the last one) and edits that comment's REST ID directly via `gh api .../issues/comments/$OWNED_ID -X PATCH`, instead of `--edit-last`. An intervening comment from another job can no longer cause a duplicate or a misdirected edit.

File: AGENTS.md [hunk 12]
```diff
@@ -374,10 +374,10 @@ procedure.
...
 ```
 workflow-policy ──► security ──► quality ──┬──► build ──► lighthouse
-                                           ├──► e2e
-                                           ├──► e2e-deep (non-blocking)
-                                           ├──► storybook
-                                           └──► vrt
+                    │                      ├──► e2e
+                    ▼                      ├──► e2e-deep (non-blocking)
+                  pr-size (pull_request     ├──► storybook
+                  only) ──► ci-success      └──► vrt
```
[the remark this answers is not in the document]
Comment (qnbs) [reply AGENTS.md]: Fixed in ea58ddbc — both diagrams now show pr-size branching from workflow-policy (AGENTS.md's job graph redrawn with the branch originating at workflow-policy instead of security; docs/CI.md's fan-in list annotates the pr-size line with "(needs workflow-policy)").

File: docs/CI.md [hunk 13]
```diff
@@ -131,6 +133,7 @@ registry gzip-decoding failure mode, while OSV failures remain blocking.
 | Job | Needs | Purpose |
 |-----|--------|---------|
 | `workflow-policy` | — | Structural (real-parser, not regex) validation of every `.github/workflows/*.yml` and `.github/actions/**/action.yml` — permissions, needs graph, SHA-pinned action references, publishing boundary (`scripts/workflow-policy-check.mjs`). Runs first, before any job invokes the governed `./.github/actions/setup` composite, using only external SHA-pinned actions directly (never that composite) and an `--ignore-scripts --ignore-pnpmfile` install, so a PR tampering with the composite (or its own install hooks) can't execute before the gate evaluates it. |
+| `pr-size` | `workflow-policy` | PR-size governance (`scripts/check-pr-size.mjs`) — tiered file/line/commit limits, advisory below the absolute ceiling and blocking only above it. `pull_request` only. Runs the **base ref's** own copy of the checker (never the PR's working-tree copy) so a PR touching it can't raise its own limits. |
```
[the remark this answers is not in the document]
Comment (qnbs) [reply docs/CI.md]: Fixed in ea58ddbc — docs/CI.md's pr-size row now states it runs the base ref's checker "when it exists there", with a one-time fallback to the PR's own copy for the introducing PR whose base ref has no checker yet.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -159,6 +159,80 @@ jobs:
...
+  # ----------------------------------------------------------
+  pr-size:
+    name: 📏 PR Size Governance
+    if: github.event_name == 'pull_request'
+    needs: [workflow-policy]
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    permissions:
+      contents: read
+      pull-requests: write
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml:174]: Confirmed real — this is a genuine trust-boundary gap, not a bounded fix for this PR. This PR is under terminal convergence / scope freeze: pulling a trusted-execution redesign in here would repeat the exact correction-loop growth #505 already went through for the structurally-similar workflow-policy gate. Deferred to #510 (which generalizes this to workflow-policy too, since needs: alone doesn't establish trust there either — related to #506 Finding C, filed separately since #506 already needs its own future consolidation pass). Also corrected this PR's description: the "self-defeat-proofing ... or disable itself" claim overstated the guarantee — it now says explicitly this covers the checker script's logic/thresholds only, not a complete trusted-execution boundary for the job itself.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -159,6 +159,80 @@ jobs:
...
+          { printf '%s\n' "$MARKER"; cat /tmp/pr-size-report.txt; } > /tmp/pr-size-marked.txt
+          if [ "$LAST_IS_OURS" = "true" ]; then
+            gh pr comment "$PR" --edit-last --body-file /tmp/pr-size-marked.txt || true
+          else
+            gh pr comment "$PR" --body-file /tmp/pr-size-marked.txt || true
+          fi
+
+      - name: Fail only on an absolute-ceiling violation
+        if: steps.size.outputs.exit_code == '1'
+        run: exit 1
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in ea58ddbc — the fail-step condition changed from `steps.size.outputs.exit_code == '1'` to `!= '0'`, so a crash/OOM-kill/any other nonzero exit now fails the job instead of silently passing unmeasured. The script itself only ever sets exit code 0 or 1, so this doesn't change deliberate-blocking semantics.

File: scripts/check-pr-size.mjs [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+export function evaluatePrSize(base, head, dependencies = {}) {
+  const numstat = getChangedFilesNumstat(base, head, dependencies);
+  const commitCount = getCommitCount(base, head, dependencies);
+  if (numstat === null || commitCount === null) {
+    return { ok: false, error: 'could not resolve diff/commit range via git' };
+  }
+  const rows = parseNumstat(numstat);
+  const fileCount = rows.length;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Fixed in ea58ddbc — added computeGovernedFileCount(), applying the same generated-artifact/pnpm-lock.yaml exclusions already used for meaningful-line counting to the file-count ceiling too. A 19-locale atomic parity edit (19 source files + 19 rebuilt bundles = 38 total) now measures as 19 governed files, well under the 30-file absolute ceiling — the report still surfaces the true total via a "(N total incl. generated)" note. Covered by a new integration test reproducing exactly this scenario.

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -159,6 +159,81 @@ jobs:
...
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          PR="${{ github.event.pull_request.number }}"
+          REPO="${{ github.repository }}"
+          MARKER="<!-- pr-size-governance -->"
+          # QNBS-v3: find OUR comment by marker, anywhere in history — not just whichever bot comment is last.
+          OWNED_ID=$(gh api "repos/$REPO/issues/$PR/comments" \
+            --jq "[.[] | select(.user.login == \"github-actions[bot]\" and (.body | contains(\"$MARKER\")))] | last | .id // empty" \
+            2>/dev/null || echo "")
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Fixed in 1bb81035 — added --paginate to the gh api call so a marker-bearing comment past page 1 is still found on a long-running PR.

File: scripts/check-pr-size.mjs [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+      const newPath = tokens[i + 2] ?? '';
+      rows.push({ path: newPath, added, removed });
+      i += 3;
+    }
+  }
+  return rows;
+}
+
+// QNBS-v3: only rebuilt public/ bundles are generated — locales/**/community-templates/** source content still counts.
+const GENERATED_ARTIFACT_ROOTS = ['public/locales/', 'public/community-templates/'];
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Confirmed and fixed in 1bb81035 — verified against the actual code (communityTemplateService.ts fetches index.${locale}.json directly at runtime; content-guard.mjs only mirrors the unlocalized index.json). Narrowed the exclusion from the whole public/community-templates/ directory to the exact index.json path; index.<locale>.json variants are now governed like any other hand-authored file. Covered by two new tests (lines + file-count) proving index.json stays excluded while index.de.json counts.

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -159,6 +159,81 @@ jobs:
...
+        continue-on-error: true
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          PR="${{ github.event.pull_request.number }}"
+          REPO="${{ github.repository }}"
+          MARKER="<!-- pr-size-governance -->"
+          # QNBS-v3: --paginate concatenates every page — a marker comment past page 1 must still be found.
+          OWNED_ID=$(gh api --paginate "repos/$REPO/issues/$PR/comments" \
+            --jq "[.[] | select(.user.login == \"github-actions[bot]\" and (.body | contains(\"$MARKER\")))] | last | .id // empty" \
```
[the remark this answers is not in the document]
Comment (qnbs) [reply .github/workflows/ci.yml]: Confirmed and fixed in 6de893fa. Verified empirically that gh api --paginate --jq applies the filter per page, not on the combined result (a forced 3-page test with a simple aggregate filter returned 3 separate lines, not 1). Switched to piping --paginate's raw output through external jq -s (slurp), which combines all pages into one array before the select/last filter runs — re-verified the same test now returns exactly one ID.

File: scripts/check-pr-size.mjs [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+  const writeFile = dependencies.writeFileSync ?? writeFileSync;
+  const exists = dependencies.existsSync ?? existsSync;
+  const unlink = dependencies.unlinkSync ?? unlinkSync;
+  const attrPath = resolveInfoAttributesPath(dependencies);
+  if (!attrPath) return fn();
+  const hadFile = exists(attrPath);
+  const original = hadFile ? readFile(attrPath, 'utf8') : '';
+  try {
+    const separator = original && !original.endsWith('\n') ? '\n' : '';
+    writeFile(attrPath, `${original}${separator}* diff\n`);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Confirmed and fixed in 6de893fa. Reproduced the exact scenario in a scratch repo (a binary file plus a text file hidden via -diff): the bare "* diff" override force-texts every path, so the binary showed fake 250/247 line counts instead of the -/- binary marker. Switched to "* !diff" (unspecify rather than force-true) — re-verified in the same scratch repo that this still reveals the hidden text change while the binary correctly stays -/-. New regression test in checkPrSize.test.ts covers this exact case.

File: scripts/check-pr-size.mjs [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+      const newPath = tokens[i + 2] ?? '';
+      rows.push({ path: newPath, added, removed });
+      i += 3;
+    }
+  }
+  return rows;
+}
+
+// QNBS-v3: every public/locales/** file is i18n:bundle output — the whole subtree is safe to exclude.
+const GENERATED_ARTIFACT_ROOTS = ['public/locales/'];
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs]: Confirmed and fixed in 78ee6190 — verified build-i18n.mjs only ever writes <lang>/bundle.json under public/locales/, so the exclusion is now an exact pattern (public/locales/<lang>/bundle.json) instead of a directory-prefix match, mirroring the same fix already applied to community-templates. Also corrected 3 pre-existing tests that used a fictional nested path (.../writer/bundle.json) that never matched the real generator's flat output and only passed under the old, too-broad prefix match. New test proves a same-directory non-bundle file still counts.
